### PR TITLE
Store Editing Support (Fixes #1495)

### LINKED ISSUE #1495: Store enhancements for partial loading and client-side editing support
We are planning to refactor Store to better handle partial loading (e.g. from streaming processes) and client-side updates scenarios, including managed synchronization of changes to the server, maintenance of "dirty" state, and reversion of data to last loaded state.  This -- alongside with related changes to support inline grid editing -- will allow better support for applications that need to edit data.

To enable these change, we intend to make the underlying Records and RecordSet objects fully immutable.  We would also like to make the new RecordSet api a public API (for advanced users), allowing advanced applications supported access to the most efficient (immutable) representations of the data. 

One important bit is that we want to consistently use the language of "loadXXX" for methods loading clean data from server and "updateXXX" for making data dirtying changes on the client.  This is an important distinction for reasoning about all this.

**API changes:**
See #1509 for list of API changes and additions

** Related Tickets**
Related tickets that have been closed and subsumed are below.  Please review these and ensure this ticket addresses all of the usages before closing

Improved syntax for making client-side transactional updates (dirty data)
>>https://github.com/xh/hoist-react/issues/1321 
>>https://github.com/xh/hoist-react/issues/1376

API enhancement to allow Grid to get at Set representation of data
>>https://github.com/xh/hoist-react/issues/1454

### PR DESCRIPTION
**Pull request by haynesjm42**
Hoist P/R Checklist
-------------------

Review and check off the below. Items that do not apply can also be checked off to indicate they
have been considered. If unclear if a step is relevant, please leave unchecked and note in comments.

- [x] Up to date with `develop` branch as of last change.
- [x] Ready for review (or open as a draft PR / add `wip` label).
- [x] Added CHANGELOG entry (or N/A)
- [x] Reviewed for breaking changes (add `breaking-change` label + CHANGELOG if so, or N/A)
- [x] Updated doc comments / prop-types (or N/A)
- [N/A] Reviewed and tested on Mobile (or N/A)
- [x] Created Toolbox branch / PR (https://github.com/xh/toolbox/tree/recordUpdate)

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by haynesjm42** [comment]
#1495

**Comment by lbwexler** [comment]
Thanks for the great API docs.  Let's review this on Monday AM?  I'll send a calendar invite.

**Comment by haynesjm42** [comment]
This also fixes #1467 and #1453

File: data/impl/RecordSet.js [hunk 1]
```diff
@@ -176,8 +200,8 @@ export class RecordSet {
                     console.debug(`Attempted to update non-existent record: ${id}`);
                     return;
                 }
-                rec.parentId = existing.parentId;
```
Comment (amcclain) [inline data/impl/RecordSet.js]: We need to update the comment above re. assigning parentId

**Pull request merged**